### PR TITLE
ARC-2922 update copy button logic

### PR DESCRIPTION
**Pull request by anshengxu**
**What's in this PR?**

Update the button appearance logic. 
1. The user can only clicked all the required buttons to continue to next step.
2. All button states get reset when user switch to another tab, either 'i am a jenkin' or 'a jenkin admin in my team'

**Why**

Jenkins design update 1.1

**Affected issues**  

ARC-2922

**How has this been tested?**  

yarn test & manual UI testing
Screenshot 2024-03-05 at 1 30 10 pm

Screenshot 2024-03-05 at 1 30 02 pm
Screenshot 2024-03-05 at 1 30 31 pm
Screenshot 2024-03-05 at 1 30 38 pm
Screenshot 2024-03-05 at 1 30 44 pm
Screenshot 2024-03-05 at 1 30 49 pm
Screenshot 2024-03-05 at 1 30 59 pm
Screenshot 2024-03-05 at 1 31 06 pm
Screenshot 2024-03-05 at 1 31 19 pm

Screenshot 2024-03-05 at 1 29 32 pm
Screenshot 2024-03-05 at 1 29 42 pm
Screenshot 2024-03-05 at 1 29 44 pm (2)

### CODE REVIEW

**Comment by anshengxu** [comment]
PR review pause as I just talked with Sarah. Some design needs to be updated

File: app/jenkins-for-jira-ui/src/components/JenkinsSetup/JenkinsSetup.tsx [hunk 1]
```diff
@@ -47,65 +47,80 @@ import {
 } from '../../common/analytics/analytics-events';
 import { AnalyticsClient } from '../../common/analytics/analytics-client';
 import { ParamTypes } from '../../common/types';
+import { CopiedToClipboard } from '../CopiedToClipboard/CopiedToClipboard';
 
 const analyticsClient = new AnalyticsClient();
 
 type CopyProps = {
 	handleCopyToClipboard: (copyRef: React.RefObject<HTMLDivElement>, elementName?: string) => Promise<void> | void;
+	setCopyButtonStateToTrue: (copyButton: CopyButtonNameEnum) => void;
```
Comment (yvonnevictoria) [inline app/jenkins-for-jira-ui/src/components/JenkinsSetup/JenkinsSetup.tsx:56]: I would give this a more detailed name. What's the implication of the button being set to true? Gonna read the rest of the PR and come back

File: app/jenkins-for-jira-ui/src/components/JenkinsSetup/JenkinsSetup.tsx [hunk 2]
```diff
@@ -214,10 +261,17 @@ const JenkinsSetup = (): JSX.Element => {
...
+	const [isCopyAdminWebhook, setIsCopyAdminWebhook] = useState(false);
+	const [isCopyAdminSecret, setIsCopyAdminSecret] = useState(false);
+	const [isCopyNonAdminWebhook, setIsCopyNonAdminWebhook] = useState(false);
+	const [isCopyNonAdminSecret, setIsCopyNonAdminSecret] = useState(false);
 	const [webhookUrl, setWebhookUrl] = useState('');
 	const [secret, setSecret] = useState<string>('');
 	const [siteName, setSiteName] = useState<string>('');
 	const [globalPageUrl, setGlobalPageUrl] = useState<string>('');
+	const [primaryCopyButtonName, setPrimaryCopyButtonName] =
+	useState<CopyButtonNameEnum>(CopyButtonNameEnum.NonAdminWebhook);
```
Comment (yvonnevictoria) [inline app/jenkins-for-jira-ui/src/components/JenkinsSetup/JenkinsSetup.tsx:274]: /nit indentation

File: app/jenkins-for-jira-ui/src/components/JenkinsSetup/JenkinsSetup.tsx [hunk 3]
```diff
@@ -202,6 +241,14 @@ const IAmTheJenkinsAdmin = ({
 	);
 };
 
+const enum CopyButtonNameEnum {
```
Comment (yvonnevictoria) [inline app/jenkins-for-jira-ui/src/components/JenkinsSetup/JenkinsSetup.tsx:244]: We want to avoid using enums in ts

More info [here](https://dev.to/ivanzm123/dont-use-enums-in-typescript-they-are-very-dangerous-57bh#:~:text=Excessive%20use%20of%20regular%20ENUMs,scalability%20issues%2C%20and%20maintainability%20issues)

TLDR: they aren't typesafe, aren't as performant as objects or union/literal types and there are alternatives that are easier to work with

Could replace it with something like this:

const CopyButtonName = {
    AdminSiteName: 'adminSiteName',
    AdminWebhook: 'adminWebhook',
    AdminSecret: 'adminSecret',
    NonAdminWebhook: 'nonAdminWebhook',
    NonAdminSecret: 'nonAdminSecret'
} as const;
or

type CopyButtonName = 'adminSiteName' | 'adminWebhook' | 'adminSecret' | 'nonAdminWebhook';

File: app/jenkins-for-jira-ui/src/components/JenkinsSetup/JenkinsSetup.tsx [hunk 4]
```diff
@@ -214,10 +261,17 @@ const JenkinsSetup = (): JSX.Element => {
 	const [serverName, setServerName] = useState('');
 	const [showMyJenkinsAdmin, setShowMyJenkinsAdmin] = useState(false);
 	const [showIAmTheJenkinsAdmin, setShowIAmTheJenkinsAdmin] = useState(false);
+	const [isCopyAdminSiteName, setIsCopyAdminSiteName] = useState(false);
```
Comment (yvonnevictoria) [inline app/jenkins-for-jira-ui/src/components/JenkinsSetup/JenkinsSetup.tsx:264]: I'm not sure about the context of why you are doing this. Could you explain?

Either way, one piece of state could manage all of these if you did something like:

```
const [isCopyName, setIsCopyName] = useState({
  buttonOne: false,
  buttonTwo: false
});
```

File: app/jenkins-for-jira-ui/src/components/JenkinsSetup/JenkinsSetup.tsx [hunk 5]
```diff
@@ -252,6 +306,49 @@ const JenkinsSetup = (): JSX.Element => {
 		fetchData();
 	}, [uuid, getServer]);
 
+	const setCopiedButtonStateToTrue = (copyButtonName: CopyButtonNameEnum) => {
```
Comment (yvonnevictoria) [inline app/jenkins-for-jira-ui/src/components/JenkinsSetup/JenkinsSetup.tsx:309]: I think there's a better way to do this code. Let's have a chat over zoom :)

File: app/jenkins-for-jira-ui/src/components/JenkinsSetup/JenkinsSetup.tsx [hunk 6]
```diff
@@ -47,65 +47,80 @@ import {
...
 
 	return (
 		<div className={cx(jenkinsSetupCopyButtonContainer)} >
 			<Button
-				iconBefore={<CopyIcon label="Copy" size="medium" />}
+				iconBefore={isCopied ? <CheckIcon label="Copy" size="medium" /> : <CopyIcon label="Copy" size="medium" />}
 				onClick={() => {
 					if (copyRef && copyRef.current) {
 						handleCopyToClipboard(copyRef, testId);
 						setIsCopied(true);
```
Comment (yvonnevictoria) [inline app/jenkins-for-jira-ui/src/components/JenkinsSetup/JenkinsSetup.tsx:94]:
```
setIsCopied(true);
setShowIsCopiedClipBoard(true);
setCopyButtonStateToTrue(copyButtonName);
```

It seems that all of these variables are tracking similar things. We could probably consolidate this

Comment (anshengxu) [reply app/jenkins-for-jira-ui/src/components/JenkinsSetup/JenkinsSetup.tsx:94]: Although they are triggered by the same action, they are liable to different things. isCopied is responsible for controlling the icon and color of the button. showIsCopiedClipBoard control the clip board to be shown after clicked, which would disappear after 2 seconds. copyButtonStateToTrue set the state in its parent component. Happy to chat on your thoughts.

File: app/jenkins-for-jira-ui/src/components/JenkinsSetup/JenkinsSetup.tsx (hunk 2 above)
Comment (anshengxu) [reply app/jenkins-for-jira-ui/src/components/JenkinsSetup/JenkinsSetup.tsx:274]: Eslint would complain if I add indentation. I think eslint want this style even I prefer adding an indentation indeed.

File: app/jenkins-for-jira-ui/src/components/JenkinsSetup/JenkinsSetup.tsx (hunk 1 above)
Comment (anshengxu) [reply app/jenkins-for-jira-ui/src/components/JenkinsSetup/JenkinsSetup.tsx:56]: This is used to decide if the 'Finish' button would be disabled or enabled. A.k.a whether a user can continue to next step. If a set of the copy button states are true, then use can continue to next page. Otherwise, user will not be able to continue.

File: app/jenkins-for-jira-ui/src/components/JenkinsSetup/JenkinsSetup.tsx (hunk 4 above)
Comment (anshengxu) [reply app/jenkins-for-jira-ui/src/components/JenkinsSetup/JenkinsSetup.tsx:264]: Basically each states correspond to whether the value is copied or not, which is used to decide if user can continue to next page.

File: app/jenkins-for-jira-ui/src/components/JenkinsSetup/JenkinsSetup.tsx (hunk 6 above)
Comment (yvonnevictoria) [reply app/jenkins-for-jira-ui/src/components/JenkinsSetup/JenkinsSetup.tsx:94]: Oh sure that makes sense. Regarding the `showIsCopiedClipboard`, it would make more sense to include the word tooltip in there to differentiate it from the other functions. I would update this to `showIsCopiedToClipboardTooltip`

Ideally, the `CopyButton` component wouldn't live in this file as it should be reusable. We should move it out to a `/common` folder at a later date, which would also make this file less confusing.

File: app/jenkins-for-jira-ui/src/components/JenkinsSetup/JenkinsSetup.tsx [hunk 7]
```diff
@@ -343,6 +453,9 @@ const JenkinsSetup = (): JSX.Element => {
 
 	const isFetchingData = !serverName || !webhookUrl || !secret;
 
+	const enableFinishButton = (copyButtonStates.isCopyAdminSecret && copyButtonStates.isCopyAdminSiteName &&
```
Comment (yvonnevictoria) [inline app/jenkins-for-jira-ui/src/components/JenkinsSetup/JenkinsSetup.tsx:456]: We should rename this to `allButtonsCopied/Clicked` or something descriptive of what the conditional is checking

File: app/jenkins-for-jira-ui/src/components/JenkinsSetup/JenkinsSetup.tsx (hunk 5 above)
Comment (yvonnevictoria) [reply app/jenkins-for-jira-ui/src/components/JenkinsSetup/JenkinsSetup.tsx:312]: What you're trying to do here is setting the next primary button. I'd call a `handleCopy` function on the button click to handle all actions, as there are likely a few.  I'd also use an array to determine priority of the buttons and help you easily find the next one. The beauty of something like this is that it scales easily to add/remove buttons, and reorder priority if we want to :) 

This is how I would recommend you go about this:

```
// buttons are stored in array in priority order. this is currently what you have in your state
const buttons = [
	{ name: 'buttonOne', clicked: false },
	{ name: 'buttonTwo', clicked: false },
	{ name: 'buttonThree', clicked: false },
	{ name: 'buttonFour', clicked: false },
	{ name: 'buttonFive', clicked: false }
];

// it's nice to have a `handleXyz` function to keep your jsx tidy
const handleCopy = ({ buttonName }) => {
	// do all the things you need to do when you handle a copy here


	// decide which button is next.
	// at this point in your code, you should've set the button to be `clicked: true`
	// please note I'm returning null here if all buttons have been clicked to prevent it from being set as undefined.
	const nextPrimaryButton =  buttons.find(button => button.clicked === false) || null;
	setNextPrimaryButton(nextPrimaryButton);
}
```

More details in our slack thread!

**Review by yvonnevictoria — APPROVED** [review]
